sdk/physical/batching.go

```diff
@@ -0,0 +1,138 @@
+package physical
+
+import (
+	"context"
+	"time"
+)
+
+type batchValue struct{}
+
+var contextBatch batchValue = struct{}{}
+
+// ContextBatch yields a context that will satisfy IsBatchContext.
+func ContextBatch(ctx context.Context) context.Context {
+	return context.WithValue(ctx, contextBatch, "1")
+}
+
+// IsBatchContext indicates that with a transactional backends,
+// Put and Delete operations will be grouped into transactions.
+func IsBatchContext(ctx context.Context) bool {
+	if ctx == nil {
+		return false
+	}
+	boolRaw := ctx.Value(contextBatch)
+	return boolRaw != nil && boolRaw.(string) != ""
+}
+
+type batchRequest struct {
+	op      Operation
+	entry   Entry
+	errChan chan error
+}
+
+// A Batcher wraps a physical Backend such that if the backend supports
+// transactions, Put and Delete will batch requests into transactions
+// if they're called with a context that satisfies IsBatchContext.
+type Batcher struct {
+	timer         *time.Ticker
+	batchMax      int
+	submit        chan batchRequest
+	activeContext context.Context
+	storage       Backend
+}
+
+var _ Backend = &Batcher{}
+
+func NewBatcher(storage Backend) *Batcher {
+	return &Batcher{
+		timer:    time.NewTicker(100 * time.Millisecond),
+		batchMax: 64, // consul only allows 64
+		submit:   make(chan batchRequest),
+		storage:  storage,
+	}
+}
+
+// Start runs the Batcher service until ctx is done.  It should normally be
+// run as a goroutine.
+func (d *Batcher) Start(ctx context.Context) {
+	d.activeContext = ctx
+	var reqs []batchRequest
+	for {
+		select {
+		case <-d.activeContext.Done():
+			// Should we write to accumulated reqs errChans?
+			return
+		case req := <-d.submit:
+			reqs = append(reqs, req)
+			if len(reqs) >= d.batchMax {
```

Comment:
I think this is not what I would most want from a general batching utility. It necessarily adds up to 100ms of latency when the load is not high, which could be a big deal for an application.

I would expect to see something more like a semaphore:
 * If no Put is in flight, go ahead.
 * If a Put is currently in flight, start accumulating writes into a transaction that will occur once the previous storage action is complete.

You can increase the amount of batching by changing this slightly to:
 * If no Put has taken place in the last N milliseconds, let it through
 * Otherwise, enqueue Puts to a transaction that will commit no later than K milliseconds later

Maybe the current behavior is fine for the ExpirationManager use case though, since it is not user-visible latency, based on where the batching context is enabled?

Reply:
That's the idea, yeah.  In expiration we're fine with a slight delay, in fact it's actually a win, because it frees up some time on the storage layer to let incoming HTTP requests make use of it.

Reply:
This wasn't obvious to me, so I asked Brian in slack.  For the sake of any other readers: suppose you have 10000 leases in a tree.  For each lease, we're going to do a handful of PUTs and DELETEs.  The expiration goroutine revoking the tree is doing these operations serially, waiting for the result of each.  If there isn't enough other expiration activity going on to fill batches as fast as we can send them, revoking this giant tree will take a very long time, potentially hours.